ext/drx/drx.c

```diff
@@ -2540,97 +2556,78 @@ static bool
 drx_try_to_detect_scatter_gather_sequence(
     void *drcontext, dr_restore_state_info_t *info, instr_t *inst,
     scatter_gather_info_t *sg_info,
-    bool (*state_machine_func)(void *drcontext, byte *pc, byte *prev_pc,
-                               int *detect_state, byte **restore_dest_mask_start,
-                               byte **restore_scratch_mask_start,
-                               int *skip_unknown_instr_count, reg_id_t *the_scratch_xmm,
-                               reg_id_t *gpr_bit_mask, reg_id_t *gpr_save_scratch_mask,
-                               uint *scalar_mask_update_no, dr_restore_state_info_t *info,
-                               instr_t *inst, scatter_gather_info_t *sg_info))
+    bool (*state_machine_func)(void *drcontext, drx_state_machine_params_t *params))
 {
-    byte *prev_pc, *pc;
-    byte *restore_dest_mask_start = NULL;
-    byte *restore_scratch_mask_start = NULL;
-    int detect_state = 0;
-    int skip_unknown_instr_count = 0;
-    reg_id_t the_scratch_xmm = DR_REG_NULL;
-    reg_id_t gpr_bit_mask = DR_REG_NULL;
-    reg_id_t gpr_save_scratch_mask = DR_REG_NULL;
-    uint scalar_mask_update_no = 0;
-    pc = info->fragment_info.cache_start_pc;
+    drx_state_machine_params_t params;
+    params.restore_dest_mask_start = NULL;
+    params.restore_scratch_mask_start = NULL;
+    params.detect_state = 0;
+    params.skip_unknown_instr_count = 0;
+    params.the_scratch_xmm = DR_REG_NULL;
+    params.gpr_bit_mask = DR_REG_NULL;
+    params.gpr_save_scratch_mask = DR_REG_NULL;
+    params.scalar_mask_update_no = 0;
+    params.info = info;
+    params.sg_info = sg_info;
+    params.inst = inst;
+    params.pc = params.info->fragment_info.cache_start_pc;
     /* As the state machine is looking for blocks of code that the fault may hit, the 128
      * bytes is a conservative approximation of the block's size, see (a) and (b) above.
      */
-    while (pc <= info->raw_mcontext->pc + 128) {
-        instr_reset(drcontext, inst);
-        prev_pc = pc;
-        pc = decode(drcontext, pc, inst);
-        if (pc == NULL) {
+    while (params.pc <= params.info->raw_mcontext->pc + 128) {
+        instr_reset(drcontext, params.inst);
+        params.prev_pc = params.pc;
+        params.pc = decode(drcontext, params.pc, params.inst);
```

Comment:
This is unusual: clobbering a passed-in instr_t pointer.  The parameter should be labeled in the signature to make that clear.  Better, and more normal, would be to make a new instr_t here for use inside this function, instead of clobbering what was passed in.  Why is it passed in at all?  I don't see it used anywhere.  Labeling the param IN or INOUT or OUT may help.  But it looks like it should just be removed to me.

Reply:
Done!